File: services/agent-runtime/packages/runtime-pod/src/run-tool-canonical-json.ts
/**
 * Canonicalizes JSON tool inputs for cross-service idempotency hashing. The
 * canonical form removes insignificant whitespace and sorts object members by
 * the UTF-8 bytes of their original quoted key tokens while preserving scalar
 * token spelling and array order. It validates the complete input with `JSON.parse`, then constructs
 * canonical output from the original scalar tokens without decoding and re-encoding them. It rejects
 * nesting beyond the local bound and performs no I/O.
 *
 * `RuntimePodToolRunner` calls this module before hashing Bridge tool inputs and
 * before sending canonical JSON payloads to Bridge and MCP Connector.
 */
const RunToolCanonicalJSONMaxDepth = 256;

// Keep mechanically identical to Bridge's run_tool_canonical_json.go. Scalar
// tokens are never decoded: only insignificant whitespace and object-member
// order change.
/**
 * Produces the compact canonical representation of one JSON document.
 *
 * @param raw - A complete JSON document whose scalar token spelling must remain
 * unchanged.
 * @returns Canonical JSON with recursively sorted object members.
 * @throws `SyntaxError` for invalid JSON and `Error` for input that exceeds the
 * nesting bound or cannot be fully consumed by the canonical parser.
 */
export function canonicalRunToolJSON(raw: string): string {
  JSON.parse(raw);
  const parser = new CanonicalJSONParser(raw);
  const canonical = parser.parseValue(0);
  parser.skipWhitespace();
  if (!parser.atEnd()) {
    throw new Error("trailing JSON input");
  }
  return canonical;
}

class CanonicalJSONParser {
  private offset = 0;

  constructor(private readonly raw: string) {}

  parseValue(depth: number): string {
    if (depth > RunToolCanonicalJSONMaxDepth) {
      throw new Error("JSON nesting exceeds canonicalization bound");
    }
    this.skipWhitespace();
    const next = this.raw[this.offset];
    if (next === undefined) {
      throw new Error("missing JSON value");
    }
    if (next === "{") return this.parseObject(depth + 1);
    if (next === "[") return this.parseArray(depth + 1);
    if (next === '"') return this.parseStringToken();
    const start = this.offset;
    while (!this.atEnd() && !isJSONDelimiter(this.raw[this.offset]!)) this.offset++;
    if (start === this.offset) throw new Error("invalid JSON scalar");
    return this.raw.slice(start, this.offset);
  }

  private parseObject(depth: number): string {
    this.offset++;
    this.skipWhitespace();
    const members: Array<{ key: string; value: string; keyBytes: Uint8Array }> = [];
    if (this.consume("}")) return "{}";
    for (;;) {
      this.skipWhitespace();
      const key = this.parseStringToken();
      this.skipWhitespace();
      if (!this.consume(":")) throw new Error("missing JSON object colon");
      const value = this.parseValue(depth);
      members.push({ key, value, keyBytes: new TextEncoder().encode(key) });
      this.skipWhitespace();
      if (this.consume("}")) break;
      if (!this.consume(",")) throw new Error("missing JSON object comma");
    }
    members.sort((left, right) => compareBytes(left.keyBytes, right.keyBytes));
    return `{${members.map(({ key, value }) => `${key}:${value}`).join(",")}}`;
  }

  private parseArray(depth: number): string {
    this.offset++;
    this.skipWhitespace();
    const values: string[] = [];
    if (this.consume("]")) return "[]";
    for (;;) {
      values.push(this.parseValue(depth));
      this.skipWhitespace();
      if (this.consume("]")) break;
      if (!this.consume(",")) throw new Error("missing JSON array comma");
    }
    return `[${values.join(",")}]`;
  }

  private parseStringToken(): string {
    if (!this.consume('"')) throw new Error("missing JSON string");
    const start = this.offset - 1;
    while (!this.atEnd()) {
      const next = this.raw[this.offset]!;
      if (next === '"') {
        this.offset++;
        return this.raw.slice(start, this.offset);
      }
      if (next === "\\") {
        const escape = this.raw[this.offset + 1];
        this.offset += escape === "u" ? 6 : 2;
      } else {
        this.offset++;
      }
    }
    throw new Error("unterminated JSON string");
  }

  skipWhitespace(): void {
    while (!this.atEnd() && isJSONWhitespace(this.raw[this.offset]!)) this.offset++;
  }

  atEnd(): boolean {
    return this.offset >= this.raw.length;
  }

  private consume(want: string): boolean {
    if (this.raw[this.offset] !== want) return false;
    this.offset++;
    return true;
  }
}

function compareBytes(left: Uint8Array, right: Uint8Array): number {
  const common = Math.min(left.length, right.length);
  for (let index = 0; index < common; index++) {
    if (left[index] !== right[index]) return left[index]! - right[index]!;
  }
  return left.length - right.length;
}

function isJSONWhitespace(value: string): boolean {
  return value === " " || value === "\t" || value === "\r" || value === "\n";
}

function isJSONDelimiter(value: string): boolean {
  return isJSONWhitespace(value) || value === "," || value === "]" || value === "}";
}

File: services/bridge/run_tool_canonical_json.go
package agentruntimebridge

import (
	"bytes"
	"encoding/json"
	"errors"
	"sort"
	"strings"
)

const runToolCanonicalJSONMaxDepth = 256

// canonicalRunToolInput implements the same raw-token algorithm as Runtime's
// canonicalRunToolJSON. It never decodes JSON strings or numbers: insignificant
// whitespace is removed, arrays retain order, and object members are ordered
// by the UTF-8 bytes of their raw key token. This preserves every scalar byte
// emitted by JSON.stringify, including lone-surrogate escapes.
func canonicalRunToolInput(raw string) (string, string, error) {
	canonical, err := canonicalRunToolJSON(raw)
	if err != nil {
		return "", "", err
	}
	return canonical, sha256Hex(canonical), nil
}

func canonicalRunToolJSON(raw string) (string, error) {
	return canonicalRunToolJSONWithoutObjectFields(raw, nil)
}

func canonicalRunToolJSONWithoutObjectFields(raw string, excludedObjectFields map[string]struct{}) (string, error) {
	if !json.Valid([]byte(raw)) {
		return "", errors.New("invalid JSON")
	}
	parser := runToolCanonicalJSONParser{raw: raw, excludedObjectFields: excludedObjectFields}
	canonical, err := parser.parseValue(0)
	if err != nil {
		return "", err
	}
	parser.skipWhitespace()
	if parser.offset != len(raw) {
		return "", errors.New("trailing JSON input")
	}
	return canonical, nil
}

type runToolCanonicalJSONParser struct {
	raw                  string
	offset               int
	excludedObjectFields map[string]struct{}
}

type runToolCanonicalJSONMember struct {
	key   string
	value string
}

func (p *runToolCanonicalJSONParser) parseValue(depth int) (string, error) {
	if depth > runToolCanonicalJSONMaxDepth {
		return "", errors.New("JSON nesting exceeds canonicalization bound")
	}
	p.skipWhitespace()
	if p.offset >= len(p.raw) {
		return "", errors.New("missing JSON value")
	}
	switch p.raw[p.offset] {
	case '{':
		return p.parseObject(depth + 1)
	case '[':
		return p.parseArray(depth + 1)
	case '"':
		return p.parseStringToken()
	default:
		start := p.offset
		for p.offset < len(p.raw) && !isRunToolJSONDelimiter(p.raw[p.offset]) {
			p.offset++
		}
		if start == p.offset {
			return "", errors.New("invalid JSON scalar")
		}
		return p.raw[start:p.offset], nil
	}
}

func (p *runToolCanonicalJSONParser) parseObject(depth int) (string, error) {
	p.offset++
	p.skipWhitespace()
	members := make([]runToolCanonicalJSONMember, 0)
	if p.consume('}') {
		return "{}", nil
	}
	for {
		p.skipWhitespace()
		key, err := p.parseStringToken()
		if err != nil {
			return "", err
		}
		p.skipWhitespace()
		if !p.consume(':') {
			return "", errors.New("missing JSON object colon")
		}
		value, err := p.parseValue(depth)
		if err != nil {
			return "", err
		}
		excluded, err := p.objectFieldExcluded(key)
		if err != nil {
			return "", err
		}
		if !excluded {
			members = append(members, runToolCanonicalJSONMember{key: key, value: value})
		}
		p.skipWhitespace()
		if p.consume('}') {
			break
		}
		if !p.consume(',') {
			return "", errors.New("missing JSON object comma")
		}
	}
	sort.SliceStable(members, func(left int, right int) bool {
		return bytes.Compare([]byte(members[left].key), []byte(members[right].key)) < 0
	})
	var canonical strings.Builder
	canonical.WriteByte('{')
	for index, member := range members {
		if index > 0 {
			canonical.WriteByte(',')
		}
		canonical.WriteString(member.key)
		canonical.WriteByte(':')
		canonical.WriteString(member.value)
	}
	canonical.WriteByte('}')
	return canonical.String(), nil
}

func (p *runToolCanonicalJSONParser) objectFieldExcluded(rawKey string) (bool, error) {
	if len(p.excludedObjectFields) == 0 {
		return false, nil
	}
	var key string
	if err := json.Unmarshal([]byte(rawKey), &key); err != nil {
		return false, err
	}
	_, excluded := p.excludedObjectFields[key]
	return excluded, nil
}

func (p *runToolCanonicalJSONParser) parseArray(depth int) (string, error) {
	p.offset++
	p.skipWhitespace()
	values := make([]string, 0)
	if p.consume(']') {
		return "[]", nil
	}
	for {
		value, err := p.parseValue(depth)
		if err != nil {
			return "", err
		}
		values = append(values, value)
		p.skipWhitespace()
		if p.consume(']') {
			break
		}
		if !p.consume(',') {
			return "", errors.New("missing JSON array comma")
		}
	}
	return "[" + strings.Join(values, ",") + "]", nil
}

func (p *runToolCanonicalJSONParser) parseStringToken() (string, error) {
	if !p.consume('"') {
		return "", errors.New("missing JSON string")
	}
	start := p.offset - 1
	for p.offset < len(p.raw) {
		switch p.raw[p.offset] {
		case '"':
			p.offset++
			return p.raw[start:p.offset], nil
		case '\\':
			p.offset += 2
			if p.offset <= len(p.raw) && p.raw[p.offset-1] == 'u' {
				p.offset += 4
			}
		default:
			p.offset++
		}
	}
	return "", errors.New("unterminated JSON string")
}

func (p *runToolCanonicalJSONParser) skipWhitespace() {
	for p.offset < len(p.raw) {
		switch p.raw[p.offset] {
		case ' ', '\t', '\r', '\n':
			p.offset++
		default:
			return
		}
	}
}

func (p *runToolCanonicalJSONParser) consume(want byte) bool {
	if p.offset < len(p.raw) && p.raw[p.offset] == want {
		p.offset++
		return true
	}
	return false
}

func isRunToolJSONDelimiter(value byte) bool {
	switch value {
	case ' ', '\t', '\r', '\n', ',', ']', '}':
		return true
	default:
		return false
	}
}
